### PR TITLE
fix #1065

### LINKED ISSUE #1065: @MaxRows has no default value (parameter usage)
Hi,

I'm trying to use `@MaxRow` for a method parameter of a `SqlObject` as indicated in the documentation:

```
public interface SampleDAO {

    @SqlQuery("SELECT * FROM samples ORDER BY TIMESTAMP DESC")
    @RegisterBeanMapper(Sample.class)
    List<Sample> findLastEntries(@MaxRows int maxRows);
}
```

Due to the lack of a default value Eclipse is giving me an error about the missing value of `@MaxRows`. If I provide a value Eclipse is happy, and the parameter `int maxRows` is used, not the value. It's no big deal, but it would be nicer if a default value exists.

### PR DESCRIPTION
**Pull request by TheRealMarnes**
#1065

### CODE REVIEW

**Comment by TheRealMarnes** [comment]
Lol I should have run a local build first. Some of the tests in jdbi itself use MaxRows wrong

**Comment by stevenschlansker** [comment]
Might also add a release note while you're at it.

**Comment by TheRealMarnes** [comment]
> [INFO] There are 150 errors reported by Checkstyle 6.16.1 with checkstyle/checkstyle-basepom.xml ruleset.
> [ERROR] src\main\java\org\jdbi\v3\core\argument\AbstractArgumentFactory.java:[0] (misc) NewlineAtEndOfFile: File does not end with a newline.
> [ERROR] src\main\java\org\jdbi\v3\core\argument\Argument.java:[0] (misc) NewlineAtEndOfFile: File does not end with a newline.
> [ERROR] src\main\java\org\jdbi\v3\core\argument\ArgumentFactory.java:[0] (misc) NewlineAtEndOfFile: File does not end with a newline.

This is why I don't run local builds yet... I hope that other PR gets accepted soon

**Comment by TheRealMarnes** [comment]
Should there maybe IAEs for nonsense maxRows values (0 and lower)?

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/customizer/internal/MaxRowsFactory.java [hunk 1]
```diff
@@ -26,19 +26,44 @@
 
 public class MaxRowsFactory implements SqlStatementCustomizerFactory
 {
+    // arbitrary number to avoid coincidences
+    public static final int DEFAULT_MAX_ROWS = -532778771;
```
Comment (qualidafial) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/customizer/internal/MaxRowsFactory.java]: `-1` seems like the typical sentinel value for this use case.

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/customizer/internal/MaxRowsFactory.java [hunk 2]
```diff
@@ -26,19 +26,44 @@
...
+
     @Override
     public SqlStatementCustomizer createForType(Annotation annotation, Class<?> sqlObjectType)
     {
-        final int maxRows = ((MaxRows)annotation).value();
-        return stmt -> ((Query)stmt).setMaxRows(maxRows);
+        throw new UnsupportedOperationException("@" + MaxRows.class.getSimpleName() + " cannot be used as a class annotation (on " + sqlObjectType.getName() + ")");
```
[the remark this answers is not in the document]
Comment (qualidafial) [reply sqlobject/src/main/java/org/jdbi/v3/sqlobject/customizer/internal/MaxRowsFactory.java]: SqlStatementCustomer defines a default method that throws an exception if you don't override it.

**Comment by TheRealMarnes** [comment]
Hm, strange. If I remove the UOE method body and let the parent method body handle, I don't get the exception.

**Comment by TheRealMarnes** [comment]
Lol, my confusion came from using the classic -1 somewhere, which is specifically handled as default value instead of wrong. This is why I used the large random number. Can we just keep that number?

**Comment by TheRealMarnes** [comment]
I'd like to leave the randon number as default value to be able to distinguish the default "ignore me" value from haphazard nonsense values when MaxRows is used on a param. The odds of bad code leading to -1 being passed as a MaxRows value seems a lot higher to me than a random huge number.

**Comment by qualidafial** [comment]
I understand the intention, but `-1` is already a widely used primitive equivalent for "no value" (or `null`).

`@BatchChunkSize.value()` has a default of `-1`, for the same use case.

**Comment by qualidafial** [comment]
Thanks for the PR! Merging.

**Pull request merged**